### PR TITLE
Refactor: risk factor model definitions

### LINKED ISSUE #130: Merge risk-factor model classes and model definition classes.
Right now, `EnergyBalanceHierarchicalModel`, `StaticHierarchicalLinearModel`, etcetera all subclass `HierarchicalLinearModel`, but their respective definition classes `LiteHierarchicalModelDefinition`, `HierarchicalLinearModelDefinition` do not. This makes it messy when registering new models and model definitions with the risk factor module factory.

To simplify, we should merge the model definitions into their respective model classes.

Side note: consider renaming `HierarchicalLinearModel` to something more general, since these models are not necessarily HLMs any more. In fact, a lot of the above could be made more readable by renaming consistently.

### PR DESCRIPTION
**Pull request by jamesturner246**
This change refactors the risk factor model definitions to inherit from a single class `RiskFactorModelDefinition`, which allows model definitions (and subsequently risk factor models) to be loaded and created in a less fragile and redundant polymorphic way. It further allows additional RF models to more easily be added, since code is now model and definition agnostic.

Rather than merging models and definitions, as was the original plan, each 'definition' object is now a 'factory' object, which can be polymporphically called to create models of appropriate type via their `create_model` method, i.e. called in `riskfactor.cpp`. 

There's probably room for more refactoring here (even further merging model and definition functionality), but this is a good first step.

Fixes #130

### CODE REVIEW

**Review by alexdewar — CHANGES_REQUESTED** [review]
This is great! That code path sorely needed a refactor and this a much nicer solution. And it's always nice to see a PR that net-removes code 😄 .

I've made some minor suggestions throughout, but the general thrust of what you've done is very sensible.

Maybe for a future PR, we could get rid of the various `load_*_model_definition` functions altogether and just make them into constructors for the various definitions instead? Then you could just do ``std::make_shared<MyModelDefinition>(parsed_json)`` directly for each of the definition types.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
+    const std::map<core::IntegerInterval, AgeGroupGenderEquation> &equations,
+    const std::map<core::Identifier, core::Identifier> &variables, double boundary_percentage)
+    : equations_{equations}, variables_{variables}, boundary_percentage_{boundary_percentage} {}
 
 HierarchicalModelType EnergyBalanceHierarchicalModel::type() const noexcept {
     return HierarchicalModelType::Dynamic;
 }
 
-const std::string &EnergyBalanceHierarchicalModel::name() const noexcept { return name_; }
+const std::string EnergyBalanceHierarchicalModel::name() const noexcept { return "Dynamic"; }
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:15]: Won't a reference to const string be optimised away anyway?

File: src/HealthGPS/energy_balance_hierarchical_model.cpp [hunk 2]
```diff
@@ -37,15 +37,30 @@ void EnergyBalanceHierarchicalModel::update_risk_factors(RuntimeContext &context
...
+        if (entry.first.contains(age)) {
+            return entry.second;
+        }
+    }
+
+    if (age < all_eqns.begin()->first.lower()) {
+        return all_eqns.begin()->second;
+    }
+
+    return all_eqns.rbegin()->second;
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:61]: This, I believe, is to handle the case where there is no exact equation, but age is higher than the provided equation age ranges, in which case it will return the equation for the highest age range.

File: src/HealthGPS/energy_balance_hierarchical_model.h [hunk 3]
```diff
@@ -1,22 +1,52 @@
...
+/// @brief Define the age group dynamic equations data type
+struct AgeGroupGenderEquation {
+    /// @brief The reference age group
+    core::IntegerInterval age_group;
+
+    /// @brief The males dynamic equations
+    std::map<core::Identifier, FactorDynamicEquation> male{};
+
+    /// @brief The females dynamic equations
+    std::map<core::Identifier, FactorDynamicEquation> female{};
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.h:30]: Probably storing the equations in some sex-agnostic struct, right? I'll make a new issue.

File: src/HealthGPS.Console/model_parser.cpp [hunk 4]
```diff
@@ -203,14 +208,17 @@ EnergyBalanceModelDefinition load_newebm_risk_model_definition(const host::poco:
...
-    return EnergyBalanceModelDefinition(std::move(nutrient_list), std::move(nutrient_equations));
+    return std::make_shared<EnergyBalanceModelDefinition>(std::move(nutrient_list),
+                                                          std::move(nutrient_equations));
 }
 
 void register_risk_factor_model_definitions(CachedRepository &repository, const ModellingInfo &info,
                                             const SettingsInfo &settings) {
     MEASURE_FUNCTION();
+    HierarchicalModelType model_type;
+    std::shared_ptr<RiskFactorModelDefinition> model_definition;
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Console/model_parser.cpp:219]: In this case, they're eventually stashed in a vector of shared_ptr base models, to allow polymorphic model definition storage in `CachedRepository`. I can't use `unique_ptr` in a vector without having to `std::move` the definition out of it later, defeating the point of the cache. I mostly agree that the cache is probably unnecessary, and should be removed in future though. I suppose I could start as a `unique_ptr` and then `std::move` into the cache vector once it's done.

File: src/HealthGPS/hierarchical_model_static.h [hunk 5]
```diff
@@ -1,33 +1,105 @@
...
+
+    /// @brief Associated p-value
+    double pvalue{};
+
+    /// @brief Associated t-value
+    double tvalue{};
+
+    /// @brief Associated standard error
+    double std_error{};
+};
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/hierarchical_model_static.h:21]: Agree. Missing data is should be fatal. I plan to do a big input data sanitisation PR in future, but didn't want to touch it just yet, lest something breaks.

File: src/HealthGPS/hierarchical_model_static.h [hunk 6]
```diff
@@ -1,33 +1,105 @@
...
+/// @brief Defines a linear regression model data type
+struct LinearModel {
+    /// @brief The model coefficients
+    std::unordered_map<core::Identifier, Coefficient> coefficients;
+
+    /// @brief The residuals standard deviation value
+    double residuals_standard_deviation;
+
+    /// @brief The R squared value
+    double rsquared{};
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/hierarchical_model_static.h:32]: As above. Will do a big input sanitisation in future.

File: src/HealthGPS/energy_balance_hierarchical_model.h [hunk 7]
```diff
@@ -25,8 +55,12 @@ class EnergyBalanceHierarchicalModel final : public HierarchicalLinearModel {
     void update_risk_factors(RuntimeContext &context) override;
 
   private:
-    std::reference_wrapper<LiteHierarchicalModelDefinition> definition_;
-    std::string name_{"Dynamic"};
+    std::reference_wrapper<const std::map<core::IntegerInterval, AgeGroupGenderEquation>>
+        equations_;
+    std::reference_wrapper<const std::map<core::Identifier, core::Identifier>> variables_;
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.h:60]: You mean  like a `shared_ptr`? I think that's the only way you'd be able to do what you described.

File: src/HealthGPS/energy_balance_model.h [hunk 8]
```diff
@@ -25,11 +28,33 @@ class EnergyBalanceModel final : public HierarchicalLinearModel {
     void update_risk_factors(RuntimeContext &context) override;
 
   private:
-    std::reference_wrapper<EnergyBalanceModelDefinition> definition_;
-    std::string name_{"Dynamic"};
+    std::reference_wrapper<const std::vector<core::Identifier>> nutrient_list_;
+    std::reference_wrapper<const std::map<core::Identifier, std::map<core::Identifier, double>>>
+        nutrient_equations_;
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_model.h:33]: As above.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:15]: Oh wait, I was being daft. It shouldn't be a reference because you're creating a new ``std::string`` (from the `const char*`). My bad.

That said, if it's not a reference it doesn't need to be `const`. The caller can then decide whether they want it const or not.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp [hunk 9]
```diff
@@ -1,18 +1,18 @@
...
-#include <iostream>
 
 namespace hgps {
 
 EnergyBalanceHierarchicalModel::EnergyBalanceHierarchicalModel(
-    LiteHierarchicalModelDefinition &definition)
-    : definition_{definition} {}
+    const std::map<core::IntegerInterval, AgeGroupGenderEquation> &equations,
+    const std::map<core::Identifier, core::Identifier> &variables, double boundary_percentage)
+    : equations_{equations}, variables_{variables}, boundary_percentage_{boundary_percentage} {}
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:9]: My bad. I hadn't clocked that `equations_` and `variables_` were references themselves.
Ignore me.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:15]: Ah, true.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:61]: I see. Maybe it needs a comment to that effect? Though commenting the code is maybe a separate endeavour...

What I meant though was that if `all_eqns` is empty, then you'll get a segfault here. There should probably be another check, either in this function or in the constructor. And maybe the same for `_variables`?

File: src/HealthGPS/energy_balance_hierarchical_model.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/energy_balance_hierarchical_model.h:30]: "Sex-agnostic struct" might be my new favourite phrase :wink:.

I'm not exactly what the right solution will look like, but it doesn't feel quite right. Sex isn't the only demographic characteristic you might want to classify your data by, so it seems a bit weird to explicitly embed it everywhere and then have virtually identical sets of functions for each sex.

File: src/HealthGPS/hierarchical_model_static.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/hierarchical_model_static.h:21]: FYI `std::optional` can be handy for these sorts of situations, because it lets you explicitly check whether a variable has been assigned a value or not.

File: src/HealthGPS/hierarchical_model_static.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/hierarchical_model_static.h:32]: Cool. But in the meantime, do we want to default-initialise `residuals_standard_deviation` too for consistency?

File: src/HealthGPS.Console/model_parser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Console/model_parser.cpp:219]: Ah, I see. Well in that case I guess you will need to copy it around. Like I say, definitely not a biggie anyways.

Have you worked out what the cache is for btw? I can't really think what would need caching... Does the code reload the config file in various places or something?

File: src/HealthGPS/energy_balance_hierarchical_model.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.h:30]: True. though for the equations, they depend on just the age range and the gender only. Will think about it.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:61]: Will make sure they're checked in the Definition constructors.

**Review by alexdewar — CHANGES_REQUESTED** [review]
Looking good. There are just two small changes I think are needed, that I mentioned in my comments:

1. You don't need to return a `const std::string`
2. Check that `equations_` (and `values_`?) is non-empty before use

File: src/HealthGPS.Console/model_parser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Console/model_parser.cpp:219]: I think its original purpose was to load backend filesystem data, and cache it to avoid reloading, but it seems to have been hijacked to pass around model definitions everywhere and for the module factory (yet another reason to do away with this module factory business in HealthGPS.Console). Its use as a filesystem (or database?) cache is still valid though. Just needs a tidy-up.

File: src/HealthGPS/energy_balance_hierarchical_model.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/energy_balance_hierarchical_model.h:60]: No, I just meant a regular reference:

```c++
class MyClass:
    MyClass(const std::string &s)
      : s_{s}
    {}

    private:
    const std::string &s_;
```

File: src/HealthGPS.Console/model_parser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Console/model_parser.cpp:219]: Ahh, I see. That makes sense.

File: src/HealthGPS/energy_balance_hierarchical_model.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.h:60]: Ah yes, you're absolutely right. Changed. Thanks

**Review by alexdewar — APPROVED** [review]
LGTM!

File: src/HealthGPS/energy_balance_hierarchical_model.cpp [hunk 10]
```diff
@@ -37,15 +46,29 @@ void EnergyBalanceHierarchicalModel::update_risk_factors(RuntimeContext &context
             current_risk_factors.at(age_key) = model_age;
         }
 
-        auto &equations = definition_.get().at(model_age);
+        auto &eqns = equations_at(model_age);
```
Comment (dalonsoa) [inline src/HealthGPS/energy_balance_hierarchical_model.cpp:49]: Why do you shorten this to `eqns`? Writing `equations` in full seems more readable.

File: src/HealthGPS.Console/model_parser.cpp [hunk 11]
```diff
@@ -101,10 +102,12 @@ HierarchicalLinearModelDefinition load_static_risk_model_definition(const host::
...
     }
 
-    return HierarchicalLinearModelDefinition{std::move(models), std::move(levels)};
+    return std::make_shared<HierarchicalLinearModelDefinition>(std::move(models),
+                                                               std::move(levels));
 }
 
-LiteHierarchicalModelDefinition load_dynamic_risk_model_definition(const host::poco::json &opt) {
+std::shared_ptr<LiteHierarchicalModelDefinition>
+load_dynamic_risk_model_definition(const host::poco::json &opt) {
```
Comment (dalonsoa) [inline src/HealthGPS.Console/model_parser.cpp:110]: Can you give me a 1-line explanation of what the purpose of using `std::make_shared` and `std::shared_ptr`. It seems like an important change in the overall architecture of the models, but as I'm new to C++, I cannot grasp it by just reading the code or googling the answer.

**Review by dalonsoa — COMMENTED** [review]
I'm not the best qualify to review C++ code but, for what I can see and for your description of the PR, you have mostly moved things around so that adding new models is easier in the future, right? 

Nothing to do in this PR, but somewhere in a technical documentation there should be an explanation about how to add a new model should the need arise. What should be subclassed? In what namespace should they lived to be found? Etc.

File: src/HealthGPS.Console/model_parser.cpp (hunk 11 above)
Comment (jamesturner246) [reply src/HealthGPS.Console/model_parser.cpp:110]: I use `std::shared_ptr`s here to allow the vector of model definitions (`shared_ptr`s to base model definition class objects) to hold derived model definition class objects and polymorphically call the correct `create_model`.

Not quite a one-liner any more, but `shared_ptr` is required over a `unique_ptr` because `shared_ptr` allows us to copy it from the storage vector.

File: src/HealthGPS/energy_balance_hierarchical_model.cpp (hunk 10 above)
Comment (jamesturner246) [reply src/HealthGPS/energy_balance_hierarchical_model.cpp:49]: Not sure why that happened. I can change it back.
